app/backend/core/messagebuilder.py

```diff
@@ -16,19 +23,30 @@ class MessageBuilder:
     """
 
     def __init__(self, system_content: str, chatgpt_model: str):
-        self.messages = [{"role": "system", "content": self.normalize_content(system_content)}]
+        self.messages: list[ChatCompletionMessageParam] = [
+            ChatCompletionSystemMessageParam(role="system", content=self.normalize_content(system_content))
+        ]
         self.model = chatgpt_model
 
     def insert_message(self, role: str, content: str, index: int = 1):
         """
         Inserts a message into the conversation at the specified index,
         or at index 1 (after system message) if no index is specified.
         Args:
-            role (str): The role of the message sender (either "user" or "system").
+            role (str): The role of the message sender (either "user", "system", or "assistant").
             content (str): The content of the message.
             index (int): The index at which to insert the message.
         """
-        self.messages.insert(index, {"role": role, "content": self.normalize_content(content)})
+        message: ChatCompletionMessageParam
+        if role == "user":
```

Comment:
In theory this could be a dict with lookup based on role but I think it'd be about the same level as code and perhaps more confusing than what you have. I keep thinking that MessageBuilder is too complex and could be simplified but haven't wrapped my head around how.
So this works, just sharing my thoughts.